### PR TITLE
fix(uplink): Ensure string input for b64enc in secret.yaml

### LINKED ISSUE #664: Fix `uplink` chart issue: `invalid value; expected string`
`helm template uplink` output:

> Error: template: uplink/templates/secret.yaml:8:46: executing "uplink/templates/secret.yaml" at <b64enc>: invalid value; expected string

Unit tests should be added, doc: https://github.com/helm-unittest/helm-unittest/blob/main/DOCUMENT.md

### PR DESCRIPTION
**Pull request by Sathwik-git**
This PR fixes the uplink chart issue and added unit tests
Fixes #664

### CODE REVIEW

File: charts/uplink/templates/tests/secret_test.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+suite: test secret.yaml
+templates:
+  - secret.yaml
+tests:
+  - it: should ensure the secret has correct metadata
+    set:
+      postgresql.url: "example-url"
+      client.id: "example-id"
+      client.secret: "example-secret"
+      sentry.dsn: "example-dsn"
```
Comment (dduportal) [inline charts/uplink/templates/tests/secret_test.yaml:10]:
```suggestion
```

The problem described in #664 is caused when the chart is installed without any additional values other than the defaults (e.g. what is in the `charts/uplink/values.yaml`).

It means that the unit test, in order to simulate this behavior, must have a default nominal case with no values.

You might want  at looking on other chart's tests (ex. https://github.com/jenkins-infra/helm-charts/tree/main/charts/rsyncd/tests) , we tend to have 2 unit test suites:

- `defaults_test.yaml` which simulates a case when calling the chart only with default values
- `custom_values_test.yaml` which simulates a case when passing a set of custom values to verify how they are used

File: charts/uplink/templates/tests/secret_test.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    asserts:
+      - isKind:
+          of: Secret
+      - matchRegex:
+          path: metadata.name
+          pattern: "^uplink.*" # Assumes the "uplink.fullname" template results in names starting with "uplink"
```
Comment (dduportal) [inline charts/uplink/templates/tests/secret_test.yaml:16]:
```suggestion
      - equal:
          path: metadata.name
          value: "RELEASE-NAME-uplink."
```

We want to verify the exact behavior of the chart so a regex is not a good enough validation.
If you want to verify the naming, you have to chek the https://github.com/helm-unittest/helm-unittest documentation to see how it behaves (or look on what is the result of your test if it fails)

**Review by dduportal — CHANGES_REQUESTED** [review]
Hi, thanks for this proposal!

It is a good direction but it still requires some work:

- Please set up a proper name for this PR. THe sentence `What does this PR do?`  does not indicates the purpose of this pull request. You might want to use https://www.conventionalcommits.org/en/v1.0.0/ as a tool to help you phrase the title

- You are not solving the problem described in #664 . If you follow the issue instruction (running the command `helm template` for the `uplink` chart) you will see that your change does not fix the error message:

  ```shell
  helm-charts git:(Sathwik-git/main)$ helm template charts/uplink
  Error: template: uplink/templates/secret.yaml:8:47: executing "uplink/templates/secret.yaml" at <b64enc>: invalid value; expected string
  ```

  I've added a suggestion about your unit test explaining how to simulate this behavior in the tests to give you a direction.

**Comment by Sathwik-git** [comment]
I hope it works now

**Review by Sathwik-git — COMMENTED** [review]
changes made as per the request

**Comment by dduportal** [comment]
For info, it looks like this change failed to be deployed in the jenkins-infra public cluster with the following error:

```
in clusters/publick8s.yaml: command "/usr/local/bin/helm" exited with non-zero status:

PATH:
  /usr/local/bin/helm

ARGS:
  0: helm (4 bytes)
  1: diff (4 bytes)
  2: upgrade (7 bytes)
  3: --allow-unreleased (18 bytes)
  4: uplink (6 bytes)
  5: jenkins-infra/uplink (20 bytes)
  6: --version (9 bytes)
  7: 0.1.3 (5 bytes)
  8: --namespace (11 bytes)
  9: uplink (6 bytes)
  10: --values (8 bytes)
  11: /tmp/helmfile1465312871/uplink-uplink-values-6cf6f995bf (55 bytes)
  12: --values (8 bytes)
  13: /tmp/helmfile1103962376/uplink-uplink-values-6b6c489b89 (55 bytes)
  14: --suppress-secrets (18 bytes)
  15: --context (9 bytes)
  16: 2 (1 bytes)
  17: --reset-values (14 bytes)

ERROR:
  exit status 1

EXIT STATUS
  1

STDERR:
  Error: Failed to render chart: exit status 1: Error: unable to build kubernetes objects from release manifest: error validating "": error validating data: [apiVersion not set, kind not set]
  Error: plugin "diff" exited with error

COMBINED OUTPUT:
  Error: Failed to render chart: exit status 1: Error: unable to build kubernetes objects from release manifest: error validating "": error validating data: [apiVersion not set, kind not set]
  Error: plugin "diff" exited with error
```


The setup is visible here:

- Release (with helmfile): https://github.com/jenkins-infra/kubernetes-management/blob/dcf8ab97652775030ca18e36ad6a642207ff423e/clusters/publick8s.yaml#L168-L176
- Config (custom helm values): https://github.com/jenkins-infra/kubernetes-management/blob/main/config/uplink.yaml
- Secrets are not publicly visible (of course) but looks like the following:

```
postgresql:
    url: postgres://<redacted>@<redacted>:<redacted>@<redacted>:5432/uplink
client:
    id: <redacted>
    secret: <redacted>
sentry:
    dsn: https://<redacted>@sentry.io/<redacted>
```

**Comment by dduportal** [comment]
Hello @Sathwik-git , if you have time to help us on this problem that would be wonderful!

**Pull request merged**